pkg/planner/core/indexmerge_unfinished_path.go

```diff
@@ -368,39 +366,41 @@ func buildIntoAccessPath(
 				if path == nil {
 					continue
 				}
-				paths = []*util.AccessPath{path}
+				oneAlternative = []*util.AccessPath{path}
 			}
 			needSelection = needSelection || unfinishedPath.needKeepFilter
 
-			alternativesForORBranch = append(alternativesForORBranch, alternative{paths, needSelection})
+			if needSelection {
+				// only need to set one of the paths to true
+				oneAlternative[0].KeepIndexMergeORSourceFilter = true
+			}
+			alternativesForORBranch = append(alternativesForORBranch, oneAlternative)
 		}
 		if len(alternativesForORBranch) == 0 {
 			return nil
 		}
+		allAlternativePaths = append(allAlternativePaths, alternativesForORBranch)
+	}
 
-		bestAlternative := slices.MinFunc(
-			alternativesForORBranch, func(a, b alternative) int {
-				return cmpAlternativesByRowCount(a.paths, b.paths)
-			})
-		if len(bestAlternative.paths) == 0 {
-			return nil
-		}
-		// Succeeded to get valid path(s) for this partial path.
-		partialPaths = append(partialPaths, bestAlternative.paths...)
-		needSelectionGlobal = needSelectionGlobal || bestAlternative.needSelection
+	// 2. TODO
```

Comment:
It seems you have already handled the table filters below. So what is this TODO for?

Reply:
Oh, I should add this in the next PR.
Some logic of other code paths will be moved here.